src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,205 @@
-""" This is a dummy example """
+import requests
+import pandas as pd
+import matplotlib.pyplot as plt
+from itertools import cycle
+
+# Base URL for accessing the climate data API
+API_URL = "https://climate-api.open-meteo.com/v1/climate?"
+
+# Dictionary to store geographic coordinates for specific cities
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+
+# Define the climate metrics to fetch
+VARIABLES = "temperature_2m_mean,precipitation_sum,soil_moisture_0_to_10cm_mean"  # noqa
+MODELS = "CMCC_CM2_VHR4,FGOALS_f3_H,HiRAM_SIT_HR,MRI_AGCM3_2_S,EC_Earth3P_HR,MPI_ESM1_2_XR,NICAM16_8S"  # noqa
+VARIABLES_MODELS = [
+    f"{metric}_{model}"
+    for metric in VARIABLES.split(",")
+    for model in MODELS.split(",")
+]
+
+
+def get_data_meteo_api(city, start_year="1950-01-01", end_year="2050-12-31"):
+    """Fetch climate data for a specific city within a given timeframe."""
+    params = {
+        "latitude": COORDINATES[city]["latitude"],
+        "longitude": COORDINATES[city]["longitude"],
+        "start_date": start_year,
+        "end_date": end_year,
+        "models": MODELS,
+        "daily": VARIABLES,
+    }
+
+    try:
+        response = requests.get(API_URL, params=params)
+        response.raise_for_status()
+
+        data = response.json()
+        try:
+            validate_response_schema(data)
+            return data["daily"]
+        except Exception as e:
+            print(f"Error validating response schema: {e}")
+            return None
+
+    except requests.exceptions.HTTPError as e:
+        if response.status_code == 429:
+            print("Too many requests. Please wait a bit before trying again.")
+        else:
+            print(f"HTTP Error: {e}")
+        return None
+    except requests.exceptions.RequestException as e:
+        print(f"Request failed: {e}")
+        return None
+    except ValueError as e:
+        print(f"Error decoding JSON: {e}")
+        return None
+    except Exception as e:
+        print(f"An unexpected error occurred: {e}")
+        return None
+
+
+def validate_response_schema(data):
+    if "daily" not in data:
+        raise Exception("Daily data not found in response.")
+
+    base_metrics = [metric.split("_")[0] for metric in VARIABLES.split(",")]
+    daily_keys = data["daily"].keys()
+
+    for base_metric in base_metrics:
+        if not any(key.startswith(base_metric) for key in daily_keys):
+            raise Exception(f"Metric {base_metric} not found in response.")
+
+    if "time" not in data["daily"]:
+        raise Exception("Time data not found in response.")
+
+    if len(data["daily"]["time"]) == 0:
+        raise Exception("No time data found in response.")
+
+
+def process_data(data):
```

Comment:
This is a very long function. Can you think of a way of splitting it into smaller chunks?